numpy/random/mtrand/mtrand.pyx

```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             x_ptr = <char*><size_t>x.ctypes.data
             stride = x.strides[0]
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```

Comment:
Yes there is a bug for the object case

    a = np.array([[np.arange(1), np.arange(53)], np.arange(241)])

    for _ in range(100):
        z = np.random.permutation(a)

Will segfault in both the old and new versions.

You're right in saying that the buffer is the right size for an object pointer in this case, so that can't be the cause.